File: MergeIntervals.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * Created by yhf on 15/6/13.
 */

/*
Given a collection of intervals, merge all overlapping intervals.

For example,
Given [1,3],[2,6],[8,10],[15,18],
return [1,6],[8,10],[15,18].
*/

class Interval {
    int start;
    int end;
    Interval() { start = 0; end = 0; }
    Interval(int s, int e) { start = s; end = e; }
}

class IntervalComparator implements Comparator<Interval> {
    public int compare(Interval i1, Interval i2) {
        return i1.start - i2.start;
    }
}

public class MergeIntervals {
    public static List<Interval> merge(List<Interval> intervals) {
        if (intervals.size() == 0) {
            return intervals;
        }

        // sort first
        Collections.sort(intervals, new IntervalComparator());

        // then merge
        for (int i = 1; i < intervals.size(); ) {
            if (intervals.get(i).start <= intervals.get(i - 1).end) {
                if (intervals.get(i).end < intervals.get(i - 1).end) {
                } else {
                    intervals.get(i - 1).end = intervals.get(i).end;
                }
                intervals.remove(i);
            } else {
                i++;
            }
        }
        return intervals;
    }

    public static void main(String[] args) {
        Interval i1 = new Interval(15, 18);
        Interval i2 = new Interval(8, 10);
        Interval i3 = new Interval(2, 5);
        Interval i4 = new Interval(1, 4);
        List<Interval> intervals = new ArrayList<>();
        intervals.add(i1);
        intervals.add(i2);
        intervals.add(i3);
        intervals.add(i4);
        List<Interval> res = merge(intervals);
    }
}

File: Leetcode-Python/MergeIntervals.py
"""
Given a collection of intervals, merge all overlapping intervals.

For example,
Given [1,3],[2,6],[8,10],[15,18],
return [1,6],[8,10],[15,18].

[1,12], [2,6], [7,8], [9,10], [11-14]
[1,12]
"""


class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

    def __str__(self):
        return "Interval [%s]" % str((self.start, self.end))

    def __repr__(self):
        return "[%s, %s]" % (self.start, self.end)


def cmp(i1, i2):
    return i1.start.__cmp__(i2.start) * 2 + i1.end.__cmp__(i2.end)

Interval.__cmp__ = cmp


class Solution:
    # @param {Interval[]} intervals
    # @return {Interval[]}
    def merge(self, intervals):
        if not intervals:
            return []
        intervals = sorted(intervals)
        merged = []

        pre = intervals[0]
        for interval in intervals[1:]:
            if interval.start <= pre.end:
                pre.end = max(pre.end, interval.end)
            else:
                merged.append(pre)
                pre = interval

        merged.append(pre)
        return merged


if __name__ == '__main__':
    # intervals = [Interval(2, 18), Interval(8, 10), Interval(2, 6), Interval(1, 3)]
    intervals = [Interval(1, 4), Interval(2, 5)]
    print Solution().merge(intervals)
